Snakemake dl experiments (#634)
usgs/wdfn-blog: content/snakemake-ml-experiments.md

> 

Author: ```suggestion
```

- the `prep_io_data` rule to prepare our data for model training 
- the `train_model` rule that used the prepared data (the output of the `prep_io_data` rule) to train the DL model
- and the `make_predictions` rule that took the trained model and the prepared data to make predictions

With these rules Snakemake could make a directed acyclic graph (DAG) that mapped the entire process of getting the end products (a csv file with the performance metrics for each model run) from the initial inputs (the input and observation files). 
 

## Wildcards, `expand`ing, and gathering: scaling to 27,270 branches
The prep_io_data rule was performed for each of the 101 sites and was the same for each random starting point and multi-task scaling factor.
The train, predict, and evaluate rules, were applied to all 27,270 model runs.
These processes were the same for every `observation site/multi-task scaling factor/random initialization` combination.
Snakemake's "wildcard" and "expand" functionalities allowed us to apply the same rules to all these combinations and keep track of them all.
We also defined a rule to gather all of the metric files for the final analysis.

> ### Wildcards

Author: I've tried to make this clearer.

> ### Wildcards

Author: I've tried to make this clearer.

Wildcards in Snakemake are placeholders in the output file definitions indicated by curly braces "{}".
These wildcards tell Snakemake how to parse key information from filepaths given.
For example, the `evaluate_predictions` rule (above) had wildcards for `site_id`, `factor` (the multi-task scaling factor), `seed`, and `partition`.
With these wildcards, Snakemake could use the output file path to get the correct values to pass to the `calculate_prediction` function.

> For example, if we asked Snakemake to produce `"expA/site_id_01022500/factor_10/rand_seed_4/validation_metrics.csv"`, it would make the following function call in the `run` step: 

Author: great idea.

> For example, if we asked Snakemake to produce `"expA/site_id_01022500/factor_10/rand_seed_4/validation_metrics.csv"`, it would make the following function call in the `run` step: 

Author: great idea.

The `expand` function makes all possible combinations of these three variables (i.e., 27,270) and runs the associated branch for each one.

### Gathering
Once the 27,270 models were trained and the predictions were made and evaluated, we needed to "gather" all of the metrics into one file so we could compare performance across multi-task scaling factors.
For this gathering we wrote a `combine_metrics` rule.

```
rule combine_metrics:
    input:
		expand("expA/site_id_{sites}/factor_{factors}/rand_seed_{seeds}/validation_metrics.csv"`,
				sites=all_sites,
				factors=all_factors,
				seeds=all_seeds)
    output:
        "expA/overall_metrics.csv"
    run:
        combine_exp_metrics(input, output[0])
```
In this rule, the `combine_exp_metrics` Python function took the list of the 27,270 csv files (produced by the `expand` function), read them into Pandas DataFrames, combined them into a single Pandas DataFrame, and wrote that DataFrame to a new csv file.

> Finally, to let Snakemake know that this was the ultimate file we needed we added this in the first rule of the Snakefile:

Author: another great idea

> Finally, to let Snakemake know that this was the ultimate file we needed we added this in the first rule of the Snakefile:

Author: another great idea

Below is our `evaluate_predictions` rule.
It defines (1) model predictions and observation files as the `input`, (2) a metrics .csv file as the `output`, and (3) the `calculate_metrics` Python function in the `run` section as the processing step to use the `input` to produce the `output`.

```
rule evaluate_predictions:
    input:
        "expA/site_id_{site_id}/factor_{factor}/rand_seed_{seed}/{partition}_preds.csv",
        "obs_file.csv"
    output:
        "expA/site_id_{site_id}/factor_{factor}/rand_seed_{seed}/{partition}_metrics.csv"
    run:
        calculate_metrics(input[0], input[1], output[0], partition=wildcards.partition)
```
The curly braces are "wildcard" placeholders which we'll talk more about later.

In addition to the `evaluate_predictions` rule, we defined 
- the `prep_io_data` rule to prepare our data for model training,
- the `train_model` rule that used the prepared data (the output of the `prep_io_data` rule) to train the DL model,
- and the `make_predictions` rule that took the trained model and the prepared data to make predictions.

## Wildcards, `expand`ing, and gathering: scaling to 27,270 branches

> The `prep_io_data` rule was performed for each of the 101 sites and was the same for each random starting point and multi-task scaling factor.

Author: ```suggestion
The `prep_io_data` rule was performed once for each of the 101 sites.
Then each site's single prepared data file site was used for each `random initialization`/`multi-task scaling factor` combination.
```

> The train, predict, and evaluate rules, were applied to all 27,270 model runs.

Author: ```suggestion
The `train`, `predict`, and `evaluate` rules, were applied to all 27,270 model runs.
```
